<img src="https://www.promocionachile.cl/img/logos/uai.JPG" width="400" alt="utfsm-logo" align="left"/>

# LEC de Probabilidad y Estadística, Viña del mar.
## Clase 01
### Profesor: Fabián Ramírez

# Objetivos:
* Instalar Software R.
* Conocer los beneficios de útilizar un software estadístico.
* Leer bases de datos.
* Filtrar Bases de Datos

# Presentación:
* **Fabián Ramírez Díaz**
* Estudiante de 5° año de Ingeniería civil matemática UTFSM.
* Programa de intercambio CRUV en el IES de la PUCV.
* Cursos de pedagogía en Preuniversitario Santa María.
* Contacto:
    - Via mail: fabian.ramirez@edu.uai.cl
    - Horario de consultas: jueves de 11:30 hasta las 12:30

# Evaluación:
Se realizaran 4 controles y un proyecto final. La nota del laboratorio vendra dada por:
$$
\text{Nota final} = \dfrac{60}{100}\left(\text{Promedio de los tres mejores controles}\right) + \dfrac{40}{100}(\text{Proyecto Final})
$$
Las fechas de los controles serán:
* **CONTROL 1**: 16 de Abril
* **CONTROL 2**: 7 de Mayo
* **CONTROL 3**: 28 de Mayo
* **CONTROL 4**: 18 de Junio

Todo **sujeto a modificaciones según feriados y contingencias nacionales**.

# ¿Qué es R? 

R es un lenguaje de programación enfocado a objetos, utilizada generalmente por gente que quiere realizar un estudio estadístico con la finalidad de tomar una decisión.

<img src="https://www.maximaformacion.es/wp-content/uploads/2018/08/Logo-R-Software-300x263.png" width="100" alt="utfsm-logo" align="center"/>

Se prefiere frente a otros lenguajes puesto que:
* Es **gratis** y de codigo libre.
* Es simple en comparación a otros lenguajes.
* Gran comunidad de programadores.

# Instalación del R

La instalación de R depende de su sistema operativo (Windows, Linux o Mac), por tanto dejare el link de descarga oficial de R y desde el sitio web ustedes instalaran el R dependiendo de su sistema:
* https://cran.dcc.uchile.cl/

Por desgracia la plataforma para programar es poco amigable. Por tanto se recomienda instalar **R-studio, ¡además del R!** el cual es una plataforma para programar en R de forma mas amigable.

<img src="https://josenelsonfigueiredo.files.wordpress.com/2019/02/rstudio-logo-flat.png" width="100" alt="utfsm-logo" align="center"/>



* https://rstudio.com/products/rstudio/download/

De la misma forma que para el R buscan el instalador adecuado para su sistema operativo.

# Conociendo el R
## Bases de datos:
Antes de utilizar el programa R es revisar la base de datos. Para ello se recomienda abrir la base de datos mediante algún programa como excel o libre office y deben fijarse que:
* El nombre de las variables no tenga espacios, por ejemplo si una variable se llama *clase social* debes renombrarla a *clase_social*, es común en programación cambiar los ' ' por '_'. En todo caso el R automaticamente cambia los espacios por '.' aún asi sigue siendo de buena costumbre.
* Ver que la base de datos este completa. No tenga datos faltantes, o errores de tipeo.
* **Imporante**. La base de datos debe ser guardada en formato *csv* de esa forma facilita al programa leer las bases de datos.

El formato *.csv* es un formato de cuardar un excel, lo que hace es crear un archivo de texto que divide las filas y columnas del excel a traves de puntos y comas. En este formato el R es capaz de leer la base de datos de una forma rápida y sin errores.

Para guardar un archivo excel en formato *.csv* cuando colocas *guardar como* en la parte inferior del nombre debes buscar formato *.csv*

## Leer bases de datos:
Para leer una base de datos en formato *.csv* la base de datos debe estar en la misma carpeta que se encuentra a la derecha del RStudio tal como lo indica la siguiente imagen:

<img src="https://scontent.fscl11-2.fna.fbcdn.net/v/t1.0-9/90335620_3369731349721190_1946742043430617088_n.jpg?_nc_cat=102&_nc_sid=730e14&_nc_eui2=AeG9SICpeVcjdoQua1sUE_rxWGb048ebQh2Dyi22kPfdOxAIqsNK2-x4TLYyJhDCJTHt0RoS-ofKhGvdBxDIo3ExmuteE8EoGQsxd88p-wcxPQ&_nc_ohc=rgqSxNG8ENEAX9omA-q&_nc_ht=scontent.fscl11-2.fna&oh=807e0ac4e98069f119b017d9ff05e08c&oe=5EA08B07" width="400" alt="utfsm-logo" align="center"/>


Luego de eso debes utilizar la función `read.csv('nombre_del_archivo_en_comillas')`. Además es usual guardar los datos en una variable que generalmente se le llama `data`. (se suele usar también `datos` o `df`).

En caso de querer entrar a una carpeta para tener mayor organización en tus codigos y trabajos deberas utilizar el comando `setwd('nombre_de_la_carpeta_en_comillas')`.

En el siguiente codigo se puede ver como se lee una base de datos:

In [1]:
data = read.csv('pokemon.csv')
data

X.  Name                   Type.1   Type.2 HP  Attack Defense Sp..Atk
1    1  Bulbasaur              Grass    Poison 45   49     49      65    
2    2  Ivysaur                Grass    Poison 60   62     63      80    
3    3  Venusaur               Grass    Poison 80   82     83     100    
4    4  Mega Venusaur          Grass    Poison 80  100    123     122    
5    5  Charmander             Fire            39   52     43      60    
6    6  Charmeleon             Fire            58   64     58      80    
7    7  Charizard              Fire     Flying 78   84     78     109    
8    8  Mega Charizard X       Fire     Dragon 78  130    111     130    
9    9  Mega Charizard Y       Fire     Flying 78  104     78     159    
10  10  Squirtle               Water           44   48     65      50    
11  11  Wartortle              Water           59   63     80      65    
12  12  Blastoise              Water           79   83    100      85    
13  13  Mega Blastoise         Water           79  103    120     135    
14  14  Caterpie               Bug             45   30     35      20    
15  15  Metapod                Bug             50   20     55      25    
16  16  Butterfree             Bug      Flying 60   45     50      90    
17  17  Weedle                 Bug      Poison 40   35     30      20    
18  18  Kakuna                 Bug      Poison 45   25     50      25    
19  19  Beedrill               Bug      Poison 65   90     40      45    
20  20  Mega Beedrill          Bug      Poison 65  150     40      15    
21  21  Pidgey                 Normal   Flying 40   45     40      35    
22  22  Pidgeotto              Normal   Flying 63   60     55      50    
23  23  Pidgeot                Normal   Flying 83   80     75      70    
24  24  Mega Pidgeot           Normal   Flying 83   80     80     135    
25  25  Rattata                Normal          30   56     35      25    
26  26  Raticate               Normal          55   81     60      50    
27  27  Spearow                Normal   Flying 40   60     30      31    
28  28  Fearow                 Normal   Flying 65   90     65      61    
29  29  Ekans                  Poison          35   60     44      40    
30  30  Arbok                  Poison          60   85     69      65    
⋮   ⋮   ⋮                      ⋮        ⋮      ⋮   ⋮      ⋮       ⋮      
771 771 Sylveon                Fairy            95  65     65     110    
772 772 Hawlucha               Fighting Flying  78  92     75      74    
773 773 Dedenne                Electric Fairy   67  58     57      81    
774 774 Carbink                Rock     Fairy   50  50    150      50    
775 775 Goomy                  Dragon           45  50     35      55    
776 776 Sliggoo                Dragon           68  75     53      83    
777 777 Goodra                 Dragon           90 100     70     110    
778 778 Klefki                 Steel    Fairy   57  80     91      80    
779 779 Phantump               Ghost    Grass   43  70     48      50    
780 780 Trevenant              Ghost    Grass   85 110     76      65    
781 781 Pumpkaboo Average Size Ghost    Grass   49  66     70      44    
782 782 Pumpkaboo Small Size   Ghost    Grass   44  66     70      44    
783 783 Pumpkaboo Large Size   Ghost    Grass   54  66     70      44    
784 784 Pumpkaboo Super Size   Ghost    Grass   59  66     70      44    
785 785 Gourgeist Average Size Ghost    Grass   65  90    122      58    
786 786 Gourgeist Small Size   Ghost    Grass   55  85    122      58    
787 787 Gourgeist Large Size   Ghost    Grass   75  95    122      58    
788 788 Gourgeist Super Size   Ghost    Grass   85 100    122      58    
789 789 Bergmite               Ice              55  69     85      32    
790 790 Avalugg                Ice              95 117    184      44    
791 791 Noibat                 Flying   Dragon  40  30     35      45    
792 792 Noivern                Flying   Dragon  85  70     80      97    
793 793 

Despues de eso la base de datos esta guardada en la variable `data` entonces cada ves que llamemos a `data` estaremos llamando a la base de datos.
## Trabajar con la base de datos
Ahora algo muy importante es saber como obtener las filas de una base de datos (también llamado data frame). Para ello se utiliza el simbolo de peso `$` con ello podemos acceder a una columna de los datos. Por ejemplo si quiero obtener solo los nombres de los pokemones deberia utilizar `data$Name`. En el siguiente codigo realizaremos dicho proceso:

In [12]:
data$Attack

[1]  49  62  82 100  52  64  84 130 104  48  63  83 103  30  20  45  35  25
 [19]  90 150  45  60  80  80  56  81  60  90  60  85  55  90  75 100  47  62
 [37]  92  57  72 102  45  70  41  76  45  70  45  80  50  65  80  70  95  55
 [55]  65  55  80  45  70  52  82  80 105  70 110  50  65  95  20  35  50  50
 [73]  80 100 130  75  90 105  40  70  80  95 120  85 100  65  75  75  35  60
 [91]  65  85 110  45  70  80 105  65  95  35  50  65  65  45  48  73 105 130
[109]  30  50  40  95  50  80 120 105  55  65  90  85 130   5  55  95 125  40
[127]  65  67  92  45  75  45 110  50  83  95 125 155 100  10 125 155  85  48
[145]  55  65  65 130  60  40  60  80 115 105 135 110  85  90 100  64  84 134
[163] 110 190 150 100  49  62  82  52  64  84  65  80 105  46  76  30  50  20
[181]  35  60  90  90  38  58  40  25  30  20  40  50  75  40  55  75  95  80
[199]  20  50 100  75  35  45  55  70  30  75  65  45  85  65  65  85  75  60
[217]  72  33  80  65  90  70  75  85 125  80 120  95 130 150  10 125 185  95
[235]  80 130  40  50  50 100  55  65 105  55  40  80  60  90  90  95  60 120
[253]  80  95  20  35  95  30  63  75  80  10  85 115  75  64  84 134 164  90
[271] 130 100  45  65  85 110  60  85 120 160  70  85 110 150  55  90  30  70
[289]  45  35  70  35  50  30  50  70  40  70 100  55  85  30  50  25  35  65
[307]  85  30  60  40 130  60  80 160  45  90  90  51  71  91  60 120  20  45
[325]  45  65  75  85  85 105  70  90 110 140  40  60 100  45  75  75  50  40
[343]  73  47  60  43  73  90 120 140  70  90  60 100 120  85  25  45  60 100
[361]  70 100  85 115  40  70 110 115 100  55  95  48  78  80 120  40  70  41
[379]  81  95 125  15  60  70  90  75 115 165  40  70  68  50 130 150  23  50
[397]  80 120  40  60  80  64 104  84  90  30  75  95 135 145  55  75 135 145
[415] 100  50  75  80 100  90 130 100 150 150 180 150 180 100 150 180  70  95
[433]  68  89 109  58  78 104  51  66  86  55  75 120  45  85  25  85  65  85
[451] 120  30  70 125 165  42  52  29  59  79  69  94  30  80  45  65 105  35
[469]  60  48  83 100  50  80  66  76 136  60 125  55  82  30  63  93  24  89
[487]  80  25   5  65  92  70  90 130 170  85  70 110 145  72 112  50  90  61
[505] 106 100  49  69  20  62  92 132 120  70  85 140 100 123  95  50  76 110
[523]  60  95 130  80 125 165  55 100  80  50  65  65  65  65  65  75 105 125
[541] 120 120  90 160 100 120  70  80 100  90 100 103 120 100  45  60  75  63
[559]  93 123  55  75 100  55  85  60  80 110  50  88  53  98  53  98  53  98
[577]  25  55  55  77 115  60 100  75 105 135  45  57  85 135  60  60  80 105
[595] 140  50  65  95 100 125  53  63 103  45  55 100  27  67  35  60  92  72
[613]  82 117  90 140  30  86  65  95  75  90  58  30  50  78 108 112 140  50
[631]  95  65 105  50  95  30  45  55  30  40  65  44  87  50  65  95  60 100
[649]  75  75 135  55  85  40  60  75  47  77  50  94  55  80 100  55  85 115
[667]  55  75  30  40  55  87 117 147  70 110  50  40  70  66  85 125 120  74
[685] 124  85 125 110  83 123  55  65  97 109  65  85 105  85  60  90 129  90
[703] 115 100 115 105 120 150 125 145 130 170 120  72  72  77 128 120  61  78
[721] 107  45  59  69  56  63  95  36  56  50  73  81  35  22  52  50  68  38
[739]  45  65  65 100  82 124  80  48  48  48  80 110 150  50  52  72  48  80
[757]  54  92  52 105  60  75  53  73  38  55  89 121  59  77  65  92  58  50
[775]  50  75 100  80  70 110  66  66  66  66  90  85  95 100  69 117  30  70
[793] 131 131 100 100 160 110 160 110

De esta forma podemos obtener una columna de la base de datos. Ahora también podriamos buscar datos que cumplan con una cierta condición. Recordemos que las condiciones en programación son:
* `==` para denotar igualdad entre 2 variables.
* `<=` para denotar que una variable es menor o igual a otra.
* `>=` para denotar que una variable es mayor o igual a otra.
* `<` para denotar que una variable es extrictamente menor que otra.
* `>` para denotar que una variable es extrictamente mayor que otra.
* `!=` para denotar que dos variables son distintas.

### Filtrar bases de datos
Imagina que en tu base de datos solo te sirven los datos que cumplan una condición particular. Para ello en R se utiliza la función `subset(data,condición)` esta te entrega una base de datos filtrada con la condición que tu impongas. Por ejemplo en nuestra base de datos buscaremos los Pokemons con más de 160 puntos de `Attack`, entonces:

In [17]:
subset(data,Generation == 5)

X.  Name                      Type.1   Type.2   HP  Attack Defense Sp..Atk
554 554 Victini                   Psychic  Fire     100 100    100     100    
555 555 Snivy                     Grass              45  45     55      45    
556 556 Servine                   Grass              60  60     75      60    
557 557 Serperior                 Grass              75  75     95      75    
558 558 Tepig                     Fire               65  63     45      45    
559 559 Pignite                   Fire     Fighting  90  93     55      70    
560 560 Emboar                    Fire     Fighting 110 123     65     100    
561 561 Oshawott                  Water              55  55     45      63    
562 562 Dewott                    Water              75  75     60      83    
563 563 Samurott                  Water              95 100     85     108    
564 564 Patrat                    Normal             45  55     39      35    
565 565 Watchog                   Normal             60  85     69      60    
566 566 Lillipup                  Normal             45  60     45      25    
567 567 Herdier                   Normal             65  80     65      35    
568 568 Stoutland                 Normal             85 110     90      45    
569 569 Purrloin                  Dark               41  50     37      50    
570 570 Liepard                   Dark               64  88     50      88    
571 571 Pansage                   Grass              50  53     48      53    
572 572 Simisage                  Grass              75  98     63      98    
573 573 Pansear                   Fire               50  53     48      53    
574 574 Simisear                  Fire               75  98     63      98    
575 575 Panpour                   Water              50  53     48      53    
576 576 Simipour                  Water              75  98     63      98    
577 577 Munna                     Psychic            76  25     45      67    
578 578 Musharna                  Psychic           116  55     85     107    
579 579 Pidove                    Normal   Flying    50  55     50      36    
580 580 Tranquill                 Normal   Flying    62  77     62      50    
581 581 Unfezant                  Normal   Flying    80 115     80      65    
582 582 Blitzle                   Electric           45  60     32      50    
583 583 Zebstrika                 Electric           75 100     63      80    
⋮   ⋮   ⋮                         ⋮        ⋮        ⋮   ⋮      ⋮       ⋮      
689 689 Rufflet                   Normal   Flying    70  83     50      37    
690 690 Braviary                  Normal   Flying   100 123     75      57    
691 691 Vullaby                   Dark     Flying    70  55     75      45    
692 692 Mandibuzz                 Dark     Flying   110  65    105      55    
693 693 Heatmor                   Fire               85  97     66     105    
694 694 Durant                    Bug      Steel     58 109    112      48    
695 695 Deino                     Dark     Dragon    52  65     50      45    
696 696 Zweilous                  Dark     Dragon    72  85     70      65    
697 697 Hydreigon                 Dark     Dragon    92 105     90     125    
698 698 Larvesta                  Bug      Fire      55  85     55      50    
699 699 Volcarona                 Bug      Fire      85  60     65     135    
700 700 Cobalion                  Steel    Fighting  91  90    129      90    
701 701 Terrakion                 Rock     Fighting  91 129     90      72    
702 702 Virizion                  Grass    Fighting  91  90     72      90    
703 703 Tornadus Incarnate Forme  Flying             79 115     70     125    
704 704 Tornadus Therian Forme    Flying             79 100     80     110    
705 705 Thundurus Incarnate Forme Electric Flying    79 115     70     125    
706 706 Thundurus Therian Forme   Electric Flying    79 105     70     145    
707 707 Reshiram                  Dragon   Fire     10

En R también existen las funciones clasicas de sumar,restar,multiplicar y dividir dadas por `+` `-` `*` `/`. En particular existen dos funciones `max` y `min` las cuales te entregan el máximo y el mínimo de un conjunto de datos. Por ejemplo si tu conjunto de datos es:
$$
\{ 1,2,2,0,3\}
$$
Entonces:

In [20]:
max(1,2,2,0,3)

[1] 1

In [5]:
min(1,2,2,0,3)

[1] 0

Combinado con la función `subset` podemos obtener el o los datos con mayor valor en una cierta variable. Por ejemplo podriamos buscar el Pokemon con mayor `Speed` de todos entonces:

In [24]:
subset(data,Attack==max(Attack))

X.  Name          Type.1  Type.2   HP  Attack Defense Sp..Atk Sp..Def Speed
164 164 Mega Mewtwo X Psychic Fighting 106 190    100     154     100     130  
    Generation Legendary
164 1          True

De esta forma obtenemos el Pokémon con mayor velocidad de todos. También podemos utilizar mas de una condición utilizando los conectores logicos **y** y **o** representados en R por `&` y `|` respectivamente. Por ejemplo podriamos buscar los pokemones con `HP` mayor que 130 **y** que `Type.1` sea `Water`.

In [23]:
subset(data,HP<50 & Type.1 == 'Grass')

X.  Name                   Type.1   Type.2   HP  Attack Defense Sp..Atk
1    1  Bulbasaur              Grass    Poison   45   49     49      65    
2    2  Ivysaur                Grass    Poison   60   62     63      80    
3    3  Venusaur               Grass    Poison   80   82     83     100    
4    4  Mega Venusaur          Grass    Poison   80  100    123     122    
5    5  Charmander             Fire              39   52     43      60    
10  10  Squirtle               Water             44   48     65      50    
14  14  Caterpie               Bug               45   30     35      20    
17  17  Weedle                 Bug      Poison   40   35     30      20    
18  18  Kakuna                 Bug      Poison   45   25     50      25    
21  21  Pidgey                 Normal   Flying   40   45     40      35    
25  25  Rattata                Normal            30   56     35      25    
27  27  Spearow                Normal   Flying   40   60     30      31    
29  29  Ekans                  Poison            35   60     44      40    
31  31  Pikachu                Electric          35   55     40      50    
38  38  Nidoran♂               Poison            46   57     40      40    
43  43  Vulpix                 Fire              38   41     40      50    
47  47  Zubat                  Poison   Flying   40   45     35      30    
49  49  Oddish                 Grass    Poison   45   50     55      75    
50  50  Gloom                  Grass    Poison   60   65     70      85    
51  51  Vileplume              Grass    Poison   75   80     85     110    
52  52  Paras                  Bug      Grass    35   70     55      45    
56  56  Diglett                Ground            10   55     25      35    
57  57  Dugtrio                Ground            35   80     50      50    
58  58  Meowth                 Normal            40   45     35      40    
62  62  Mankey                 Fighting          40   80     35      35    
66  66  Poliwag                Water             40   50     40      40    
69  69  Abra                   Psychic           25   20     15     105    
70  70  Kadabra                Psychic           40   35     30     120    
76  76  Bellsprout             Grass    Poison   50   75     35      70    
77  77  Weepinbell             Grass    Poison   65   90     50      85    
⋮   ⋮   ⋮                      ⋮        ⋮        ⋮   ⋮      ⋮       ⋮      
652 652 Foongus                Grass    Poison    69  55     45     55     
653 653 Amoonguss              Grass    Poison   114  85     70     85     
659 659 Ferroseed              Grass    Steel     44  50     91     24     
660 660 Ferrothorn             Grass    Steel     74  94    131     54     
661 661 Klink                  Steel              40  55     70     45     
664 664 Tynamo                 Electric           35  55     40     45     
672 672 Axew                   Dragon             46  87     60     30     
681 681 Mienfoo                Fighting           45  85     50     55     
686 686 Pawniard               Dark     Steel     45  85     70     40     
702 702 Virizion               Grass    Fighting  91  90     72     90     
719 719 Chespin                Grass              56  61     65     48     
720 720 Quilladin              Grass              61  78     95     56     
721 721 Chesnaught             Grass    Fighting  88 107    122     74     
722 722 Fennekin               Fire               40  45     40     62     
725 725 Froakie                Water              41  56     40     62     
728 728 Bunnelby               Normal             38  36     38     32     
730 730 Fletchling             Normal   Flying    45  50     43     40     
733 733 Scatterbug             Bug                38  35     40     27     
734 734 Spewpa                 Bug                45  22     60     27     
738 738 Flabébé                Fairy              44  38     39     61     
741 741 Skiddo                 Grass              66

Note que las variables que son textos se les llama variables strings y van entre comillas `' '`.

In [8]:
subset(data,Name == 'Torchic')

X.  Name    Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
277 277 Torchic Fire          45 60     40      70      50      45   
    Generation Legendary
277 3          False

In [25]:
subset(data,HP == max( HP ))

X.  Name    Type.1 Type.2 HP  Attack Defense Sp..Atk Sp..Def Speed
262 262 Blissey Normal        255 10     10      75      135     55   
    Generation Legendary
262 2          False

In [26]:
subset(data,Speed == 5 & Generation == 2)

X.  Name    Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
231 231 Shuckle Bug    Rock   20 10     230     10      230     5    
    Generation Legendary
231 2          False

In [35]:
data2= subset(data,Name == 'Shuckle')

In [38]:
respuesta  = subset(data,HP == max(HP))

In [40]:
respuesta$Name

[1] Blissey
800 Levels:  Abomasnow Abra Absol Accelgor ... Zygarde Half Forme